File: app/simulation/color-picker.ts
import { Color } from "./types";

export class ColorPicker {
  constructor(
    private index = 0,
    private rgbPallette: Color[] = [
      [255, 0, 0],
      [0, 255, 0],
      [0, 0, 255],
      [255, 255, 0],
      [0, 255, 255],
      [255, 0, 255],
      [255, 255, 255],
      [0, 0, 0],
    ],
    public currentColor: Color = [255, 0, 0],
    private colorDictionary: Record<string, Color> = {}
  ) {}

  public addColor(color: Color) {
    this.rgbPallette.push(color);
  }

  public next() {
    this.index++;

    if (this.index >= this.rgbPallette.length) {
      this.index = 0;
    }

    this.currentColor = this.rgbPallette[this.index];
  }

  public prev() {
    this.index--;

    if (this.index < 0) {
      this.index = this.rgbPallette.length - 1;
    }

    this.currentColor = this.rgbPallette[this.index];
  }

  public getCurrentColor() {
    return this.currentColor;
  }

  public add(key: string) {
    if (key in this.colorDictionary) {
      this.currentColor = this.colorDictionary[key];
    } else {
      this.colorDictionary[key] = this.currentColor;
    }
  }
}

File: python/color_picker.py
class ColorPicker:
    def __init__(self):
        self.idx = 0
        self.rgb_pallette = [
            (255, 0, 0),
            (0, 255, 0),
            (0, 0, 255),
            (255, 255, 0),
            (0, 255, 255),
            (255, 0, 255),
            (255, 255, 255),
            (0, 0, 0),
        ]
        self.current_color = self.rgb_pallette[self.idx]

        self.color_dict = {}

    def add_color(self, color):
        self.rgb_pallette.append(color)

    def next(self):
        self.idx += 1
        print(self.idx)
        if self.idx >= len(self.rgb_pallette):
            self.idx = 0
        self.current_color = self.rgb_pallette[self.idx]

    def prev(self):
        self.idx -= 1

        if self.idx < 0:
            self.idx = len(self.rgb_pallette) - 1
        self.current_color = self.rgb_pallette[self.idx]

    def get_current_color(self):
        return self.current_color

    def add(self, key):
        if key in self.color_dict:
            self.current_color = self.color_dict[key]
        else:
            self.color_dict[key] = self.current_color
